### PR TITLE
Factorize the code to make it easier to add new modes AND added a Juxtalinear edition mode

### PR DESCRIPTION
**Pull request by DanielC-N**
- added a Juxta mode that creates a translation burrito
- added a 'factoryCreateContent.js'  file to help factorize some code for 'createObsContent', 'createVersificationUSFM' and 'createAudioVersification' in 'saveProjetcsMeta.js'

- fixing some typos, fixing lint

### CODE REVIEW

**Comment by DanielC-N** [comment]
First thing first I've added comments, where I did (or intend to) add code. But I've annotated it with "Nicolas" and sometimes I've added some questions with it.  
If you can answer the questions, or have a comment on my annotation or if I'm doing something wrong, please let me know!

**Comment by selvabenz** [comment]
@DanielC-N **App is breaking, could you check? Could not move further.**

image

**CSS fix required: Icon is visible only on hover.**

image

**Comment by DanielC-N** [comment]
@selvabenz I forgot to mention that it is not yet intended to be build. "As is" it'll definitely not work since I'm still modifying the code as I'm asking questions. It's only on the code part that I need help.
Sorry for the misunderstanding.

ps : I'm terrible at css, how may I fix this Icon issue?

**Comment by DanielC-N** [comment]
@selvabenz I've updated the code so it can work properly

**Comment by selvabenz** [comment]
Hi @DanielC-N thanks for updating the code. It is working fine. Juxta mode creates the projects under Text Translation flavour is that what you have intended? Have you added the Juxta feature in this Pr?

image

**Comment by DanielC-N** [comment]
@selvabenz 

I've modified the code in order to fully work with Juxta files. Now a project is correctly labeled as "Juxtalinear"  (along with the flavour)

I've some juxtalinear `JSON` files for the tests here :

[PHP.json](https://github.com/bible-technology/scribe-scripture-editor/files/14258481/PHP.json)
[TIT.json](https://github.com/bible-technology/scribe-scripture-editor/files/14258482/TIT.json)

**Comment by DanielC-N** [comment]
@selvabenz 
I've updated the code and it's working properly now. You can create a juxtalinear project by importing your own usfms. You must use greek/hebrew usfm for now.. It's correctly listed as a "Juxtalinear" flavor. When you open the project it'll open the Juxtalinear Editor.
Here's the video of the demo usage : https://youtu.be/sarZnbxbgDY

Here's some usfm files that you can import to test the mode :  
[2JN.txt](https://github.com/bible-technology/scribe-scripture-editor/files/14363062/2JN.txt)
[PHP.txt](https://github.com/bible-technology/scribe-scripture-editor/files/14363063/PHP.txt)
[TIT.txt](https://github.com/bible-technology/scribe-scripture-editor/files/14363065/TIT.txt)

**Comment by DanielC-N** [comment]
I'm still working on the performances to load a file. For example : the book of revelation in usfm takes 10~15 sec to load for now.

**Comment by vipinpaul** [comment]
Our findings, we have already shared these things with Nikola on-call

- resources collection/table took a long time to load.
- Resource is showing all versions. If needed to show all versions, make a collapsable UI for each one with the latest as default. Users can load the previous versions only by expanding the single resources.
- Academy is not showing any resources.
- Star/Unstar is not working for Juxta projects
- If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files.
- Translation resources only show 4 resources, the rest of them are not showing.
- OBS Translation note, Questions, TWL, have missing / wrong resources are showing.
- Showing me the same content for me in all the books.
- Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend.
- TWL is not loading, it keeps on spinning
- It feels like the Audio resources are reloading multiple times.
- In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books
- Achieve project is not working for the text translation project, working for Juxta

**Comment by DanielC-N** [comment]
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* Star/Unstar is not working for Juxta projects. TBD 🔧
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. TBD 🔧
* Translation resources only show 4 resources, the rest of them are not showing. DONE, but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE, but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Achieve project is not working for the text translation project, working for Juxta. To discuss :question: :question:

**Comment by DanielC-N** [comment]
Bonus : https://git.door43.org/ru_gl/ru_obs-tn

I've fixed a bug here : https://github.com/Proskomma/scribe-scripture-editor/blob/7eafc20c7dc87c89b4d8f88fcd31f9dbdf1b4653/renderer/src/components/EditorPage/Reference/OBS/ObsTn.js#L86

The problem show up when the user put the path to the FOLDER instead of the name of the tsv file in `metadata.projects[0].path`

**Comment by DanielC-N** [comment]
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* Star/Unstar is not working for Juxta projects. TBD 🔧
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. DONE e554082 :white_check_mark:
* Translation resources only show 4 resources, the rest of them are not showing. DONE :white_check_mark: , but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE :white_check_mark: , but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Archive project is not working for the text translation project, working for Juxta. To discuss :question: :question:

**Comment by selvabenz** [comment]
@DanielC-N 
- Unable to create a new project with JUSTA mode on. 
image

**Comment by DanielC-N** [comment]
@selvabenz I've updated the messages given to the user! It's clearer now.

**Comment by DanielC-N** [comment]
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* Star/Unstar is not working for Juxta projects. TBD 🔧
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. DONE e554082 :white_check_mark:
* Translation resources only show 4 resources, the rest of them are not showing. DONE :white_check_mark: , but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE :white_check_mark: , but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Archive project is not working for the text translation project, working for Juxta. Solved :white_check_mark: (a0f4b91)

**Comment by DanielC-N** [comment]
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* **Star/Unstar is not working for Juxta projects. Done :white_check_mark:** ([58605c8](https://github.com/bible-technology/scribe-scripture-editor/pull/303/commits/58605c83923bf8b40dc29fc5d9182358b0c51ee2))
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. DONE e554082 :white_check_mark:
* Translation resources only show 4 resources, the rest of them are not showing. DONE :white_check_mark: , but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE :white_check_mark: , but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Archive project is not working for the text translation project, working for Juxta. Solved :white_check_mark: (a0f4b91)

File: package-lock.json
Comment (sijumoncy) [inline package-lock.json:1]: We moved away from npm and currently using yarn. can you remove the package-lock from the PR ? Only need to keep yarn lock

**Comment by DanielC-N** [comment]
@selvabenz 
I've fixed the bug I've found in my modifications => e5fd90b

You can test it yourself by creating a project, with 2 different files, then navigating between the two books in the project. Now the sentences are correctly loaded.

**Comment by selvabenz** [comment]
Unable to export Juxta projects, the app is frozen during the export. All the other flavor exports well.

image

**Comment by DanielC-N** [comment]
@selvabenz FIX for the export (and import) problem : b4d8e3c
It now works properly.

**Comment by selvabenz** [comment]
> @selvabenz FIX for the export (and import) problem : [b4d8e3c](https://github.com/bible-technology/scribe-scripture-editor/commit/b4d8e3c43cb5ec81090d0ee3d96a0e656a345945) It now works properly.

Will check and respond soon @DanielC-N

**Comment by selvabenz** [comment]
@DanielC-N i am facing a problem in the OBS mode, while the project is exported there is no project as zip checkbox visible but it is exporting the project as a zip file. 

Note: In the dev branch project as zip only assigned to text translation

**Comment by DanielC-N** [comment]
@selvabenz I've fixed the problem here : 3dda2e5

I did modify this behavior without noticing the change for obs. In the long run, I think this part of the code could use some refactorisation.

**Comment by selvabenz** [comment]
@DanielC-N Sounds good! thanks for the fix.

@vipinpaul they did not do the CSS fix on the create project button and the spinner @joelthe1 this PR can be merged.

**Comment by selvabenz** [comment]
@vipinpaul and @joelthe1 this PR can be merged. all the asked CSS fix have been made.

**Review by vipinpaul — APPROVED** [review]

**Pull request merged**